nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc

```diff
@@ -0,0 +1,608 @@
+  num_inputs = prop->num_inputs;
+  num_labels = prop->num_labels;
+  num_training_samples = prop->num_training_samples;
+  num_validation_samples = prop->num_validation_samples;
+  model_save_path = prop->model_save_path;
+  num_epochs = prop->num_epochs;
+  training_complete_cond = prop->training_complete_cond;
+  is_training_complete = FALSE;
+  total_num_samples =
+    (num_training_samples + num_validation_samples) * num_epochs;
+
+  ml_logd("num_inputs: %" PRId64 "", num_inputs);
+  ml_logd("num_labels: %" PRId64 "", num_labels);
+  ml_logd("num_training_samples: %" PRId64 "", num_training_samples);
+  ml_logd("num_validation_samples: %" PRId64 "", num_validation_samples);
+  ml_logd("num_epochs: %" PRId64 "", num_epochs);
+  ml_logd("Total number of data to be received: %" PRId64 "",
+          total_num_samples);
+  ml_logd("model_config: %s", model_config.c_str());
+  ml_logd("model_save_path: %s", model_save_path.c_str());
+  ml_logd("<leave>");
+}
+
+static int nntrainer_model_destructor(const GstTensorTrainerFramework *fw,
+                                      const GstTensorTrainerProperties *prop,
+                                      void **private_data) {
+  NNTrainer::NNTrainerTrain *nntrainer =
+    static_cast<NNTrainer::NNTrainerTrain *>(*private_data);
+  UNUSED(fw);
+  ml_logd("<called>");
+
+  if (!nntrainer)
+    return -1;
+
+  delete nntrainer;
+  *private_data = NULL;
+  ml_logd("<leave>");
+
+  return 0;
+}
+
+static int
+nntrainer_model_start_training(const GstTensorTrainerFramework *fw,
+                               const GstTensorTrainerProperties *prop,
+                               void *private_data) {
+  NNTrainer::NNTrainerTrain *nntrainer =
+    reinterpret_cast<NNTrainer::NNTrainerTrain *>(private_data);
+  UNUSED(fw);
+
+  ml_logd("<called>");
+  if (!nntrainer) {
+    ml_loge("Failed get nntrainer");
+  }
+  try {
+    std::thread train_thread(nntrainer_thread_func, nntrainer);
+    train_thread.detach();
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return -1;
+  }
+  ml_logd("<leave>");
+  return 0;
+}
+
+void NNTrainer::NNTrainerTrain::trainModel() {
+  pid_t pid = getpid();
+  pid_t tid = syscall(SYS_gettid);
+
+  ml_logd("<called>");
+  ml_logd("pid[%d], tid[%d]", pid, tid);
+
+  try {
+    model->setProperty({"epochs=" + std::to_string(num_epochs)});
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return;
+  }
+
+  try {
+    model->train();
+    training_loss = model->getTrainingLoss();
+    validation_loss = model->getValidationLoss();
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return;
+  }
+  ml_logd("training_loss: %f, validation_loss: %f", training_loss,
+          validation_loss);
+  try {
+    ml_logd("Save_model: %s", model_save_path.c_str());
+    model->save(model_save_path, ml::train::ModelFormat::MODEL_FORMAT_BIN);
```

Comment:
I think you do not need save here if you define the save_path in ini file.

Reply:
The MLOps save path is set as a property of tensor_trainer and is set up to nntrainer.
It will be written in the related property description that the save setting by ini is ignored.
Thank you.